src/Umbraco.Infrastructure/Serialization/UmbracoJsonTypeInfoResolver.cs

```diff
@@ -0,0 +1,57 @@
+using System.Collections.Concurrent;
+using System.Text.Json;
+using System.Text.Json.Serialization.Metadata;
+using Umbraco.Cms.Core.Composing;
+
+namespace Umbraco.Cms.Infrastructure.Serialization;
+
+public sealed class UmbracoJsonTypeInfoResolver : DefaultJsonTypeInfoResolver, IUmbracoJsonTypeInfoResolver
+{
+    private readonly ITypeFinder _typeFinder;
+    private readonly ConcurrentDictionary<Type, ISet<Type>> _subTypesCache = new ConcurrentDictionary<Type, ISet<Type>>();
+    private readonly ConcurrentDictionary<Type, ISet<Type>> _implementsCache = new ConcurrentDictionary<Type, ISet<Type>>();
+
+    public UmbracoJsonTypeInfoResolver(ITypeFinder typeFinder)
+    {
+        _typeFinder = typeFinder;
+    }
+
+    public IEnumerable<Type> FindSubTypes(Type type)
+    {
+        if (_subTypesCache.TryGetValue(type, out var cachedResult))
+        {
+            return cachedResult;
+        }
+
+        var result = _typeFinder.FindClassesOfType(type).ToHashSet();
+        _subTypesCache.TryAdd(type, result);
+        return result;
+    }
+
+    public override JsonTypeInfo GetTypeInfo(Type type, JsonSerializerOptions options)
+    {
+        var result = base.GetTypeInfo(type, options);
+
+        if (type.IsInterface)
+        {
+            var subTypes = FindSubTypes(type);
+
+            if (subTypes.Any())
```

Comment:
Not mad about it, but we could reduce nesting by exiting early if `subTypes.Any()` is false, the same goes for `if (type.IsInterface)`, if it's not an interface we can just exit early.